pkg/asset/cluster/aws/aws.go

```diff
@@ -95,6 +101,125 @@ func tagSharedVPCResources(ctx context.Context, clusterID string, installConfig
 	return nil
 }
 
+func tagSharedIAMRoles(ctx context.Context, clusterID string, installConfig *installconfig.InstallConfig) error {
+	iamRoles := sets.New[string]()
+	{
+		mpool := awstypes.MachinePool{}
+		mpool.Set(installConfig.Config.AWS.DefaultMachinePlatform)
+		if mp := installConfig.Config.ControlPlane; mp != nil {
+			mpool.Set(mp.Platform.AWS)
+		}
+		if len(mpool.IAMRole) > 0 {
+			iamRoles.Insert(mpool.IAMRole)
+		}
+	}
+
+	for _, compute := range installConfig.Config.Compute {
+		mpool := awstypes.MachinePool{}
+		mpool.Set(installConfig.Config.AWS.DefaultMachinePlatform)
+		mpool.Set(compute.Platform.AWS)
+		if len(mpool.IAMRole) > 0 {
+			iamRoles.Insert(mpool.IAMRole)
+		}
+	}
+
+	// If compute stanza was not defined, it will inherit from DefaultMachinePlatform later on.
+	if installConfig.Config.Compute == nil {
+		mpool := installConfig.Config.AWS.DefaultMachinePlatform
+		if mpool != nil && len(mpool.IAMRole) > 0 {
+			iamRoles.Insert(mpool.IAMRole)
+		}
+	}
+
+	if iamRoles.Len() == 0 {
+		return nil
+	}
+
+	logrus.Debugf("Tagging shared instance roles: %v", sets.List(iamRoles))
+
+	session, err := installConfig.AWS.Session(ctx)
+	if err != nil {
+		return fmt.Errorf("could not create AWS session: %w", err)
+	}
+
+	tagKey, tagValue := sharedTag(clusterID)
+
+	iamClient := iam.New(session, aws.NewConfig().WithRegion(installConfig.Config.Platform.AWS.Region))
+	for role := range iamRoles {
+		if _, err := iamClient.TagRoleWithContext(ctx, &iam.TagRoleInput{
```

Comment:
multiple clusters could end up using the same role and/or profile correct? Will this error if we attach the same information more than once?

Reply:
> multiple clusters could end up using the same role and/or profile correct?

Correct.

> Will this error if we attach the same information more than once?

It's not exactly the same because each cluster will have a different `infraID` included in the `tagKey`.